packages/react-query/src/useIsFetching.ts

```diff
@@ -1,24 +1,13 @@
 'use client'
-import * as React from 'react'
-import { notifyManager } from '@tanstack/query-core'
-
-import { useQueryClient } from './QueryClientProvider'
+import { useQueryState } from './useQueryState'
 import type { QueryClient, QueryFilters } from '@tanstack/query-core'
 
 export function useIsFetching(
   filters?: QueryFilters,
   queryClient?: QueryClient,
 ): number {
-  const client = useQueryClient(queryClient)
-  const queryCache = client.getQueryCache()
-
-  return React.useSyncExternalStore(
-    React.useCallback(
-      (onStoreChange) =>
-        queryCache.subscribe(notifyManager.batchCalls(onStoreChange)),
-      [queryCache],
-    ),
-    () => client.isFetching(filters),
-    () => client.isFetching(filters),
-  )
+  return useQueryState(
+    { filters: { ...filters, fetchStatus: 'fetching' } },
+    queryClient,
+  ).length
```

Comment:
I don't really know why the tests fail here - they really shouldn't, the implementations are pretty much the same. I tracked it down to the `queryObserverOptions` event not being fired initially with the new impl.

these are the events we're receiving from the cache subscription:

```
useQueryState observerResultsUpdated
useQueryState updated
fire observerOptionsUpdated
useQueryState observerOptionsUpdated
useQueryState observerRemoved
```

with the old impl, we get:

```
fire observerOptionsUpdated
useIsFetching observerOptionsUpdated
useIsFetching observerResultsUpdated
useIsFetching updated
fire observerOptionsUpdated
useIsFetching observerOptionsUpdated
useIsFetching observerRemoved
```

so we get one more `observerOptionsUpdated` event, at which time we are still fetching. In the new impl, the first event we get is `observerResultsUpdated`, at which point we already have data. I don't currently know why that is ...

Reply:
pff, it seems to have something to do with the ref workaround.

I refactored it back to the original implementation, step by step. Even if we store the `number` result of `client.isFetching()` in a ref, then read from the ref in `getSnapshot`, it fails, but if we directly read from `client.isFetching()` in `getSnapshot`, it works.

This is so baffling, I guess "don't read refs during render" is coming to bite us here . I tracked the difference by logging the value of the ref and reading `isFetching()` directly, and there are indeed differences:

```
getSnapshot 0 0
getSnapshot 0 0
getSnapshot 0 1
getSnapshot 0 1
getSnapshot 0 1
```

I think moving the read + structural sharing into `getSnapshot` could solve it 🤔

Reply:
fixed it: 4f000317

`useMutationState` likely needs the same change